src/lib.rs

```diff
@@ -6,70 +6,275 @@
 //!
-//! let result_four: String = converter.simplified_to_traditional(simplified_text);
-//! println!("{}", result_four == traditional_text); // --> true
+//! let result_four = simplified_to_traditional(simplified_text);
+//! assert_eq!(result_four, traditional_text);
+//! ```
+#![cfg_attr(feature = "bench", feature(test))]
 
 extern crate bincode;
 
-mod character_converter;
-pub use self::character_converter::Converter as CharacterConverter;
+use std::borrow::Cow;
+use std::collections::HashMap;
+
+use bincode::deserialize_from;
+use fst::raw::{Fst, Output};
+use once_cell::sync::Lazy;
+
+static T2S: Lazy<HashMap<String, String>> =
+    Lazy::new(|| deserialize_from(&include_bytes!("../data/t2s.profile")[..]).unwrap());
+static S2T: Lazy<HashMap<String, String>> =
+    Lazy::new(|| deserialize_from(&include_bytes!("../data/s2t.profile")[..]).unwrap());
+
+// create an fst containing all the keys
+static T2S_FST: Lazy<Fst<Vec<u8>>> = Lazy::new(|| {
+    let mut keys: Vec<_> = T2S.keys().collect();
+    keys.sort_unstable();
+    Fst::from_iter_set(keys).unwrap()
+});
+
+// create an fst containing all the keys
+static S2T_FST: Lazy<Fst<Vec<u8>>> = Lazy::new(|| {
+    let mut keys: Vec<_> = S2T.keys().collect();
+    keys.sort_unstable();
+    Fst::from_iter_set(keys).unwrap()
+});
+
+fn is_script(
+    raw: &str,
+    mapping: &HashMap<String, String>,
+    backup: &HashMap<String, String>,
+) -> bool {
+    let mut buffer = [0; 4];
+    for character in raw.chars() {
+        let c = character.encode_utf8(&mut buffer);
+        if mapping.get(c).is_none() {
+            if backup.get(c).is_some() {
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+pub fn is_traditional(raw: &str) -> bool {
+    is_script(raw, &T2S, &S2T)
+}
+
+pub fn is_simplified(raw: &str) -> bool {
+    is_script(raw, &S2T, &T2S)
+}
+
+fn convert_script<'a>(
+    raw: &'a str,
+    mapping: &HashMap<String, String>,
+    fst: &Fst<Vec<u8>>,
+) -> Cow<'a, str> {
+    let mut converted_characters: Option<String> = None;
+    let mut skip_bytes = 0;
+
+    while skip_bytes < raw.len() {
+        let tailstr = &raw[skip_bytes..];
+
+        match find_longest_prefix(fst, tailstr.as_bytes()) {
+            Some((_, length)) => {
+                let tailstr = &tailstr[..length];
+                let mapped = mapping.get(tailstr).unwrap();
+
+                converted_characters = match converted_characters.take() {
+                    Some(mut converted_characters) => {
+                        converted_characters.push_str(mapped);
+                        Some(converted_characters)
+                    }
+                    None => {
+                        if tailstr != mapped {
+                            let mut converted_characters = String::new();
+                            converted_characters.push_str(&raw[..skip_bytes]);
+                            converted_characters.push_str(mapped);
```

Comment:
Using a `String::new()` here is equivalent to the previous solution, you'll allocate for the first `push_str` and will maybe reallocate to fit the second `push_str`, please use `String::with_capacity`.